hal/src/nRF52840/exflash_hal.cpp

```
@@ -0,0 +1,739 @@
+
+    FlashLock lk;
+
+    hal_qspi_flash_type_t all_flashes[] = {
+        HAL_QSPI_FLASH_TYPE_MX25L3233F,
+        HAL_QSPI_FLASH_TYPE_MX25R6435F,
+        HAL_QSPI_FLASH_TYPE_GD25WQ64E
+    };
+
+    // Use single data line SPI to read ID
+    nrfx_qspi_config_t default_config = NRF_QSPI_QUICK_CFG(NRF_QSPI_WRITEOC_PP, NRF_QSPI_READOC_FASTREAD, NRF_QSPI_FREQ_32MDIV2);
+    int ret = nrfx_qspi_init(&default_config, nullptr, nullptr);
+    if (ret == NRFX_ERROR_TIMEOUT) {
+        // Could have timed out because the flash chip might be in an ongoing program/erase operation or Deep Power-Down mode.
+        // Suspend it or wake it up.
+        LOG_DEBUG(TRACE, "QSPI NRFX timeout error. Suspend pgm/ers op or wake it up.");
+
+        // Try all the supported suspend command
+        for (auto flash: all_flashes) {
+            flash_type = flash;
+            hal_exflash_special_command(HAL_EXFLASH_SPECIAL_SECTOR_NONE, HAL_EXFLASH_COMMAND_SUSPEND_PGMERS, nullptr, nullptr, 0);
+            nrf_delay_us(20); // Need delay to ensure the flash is suspended
+        }
+    }
+
+    // Determine chip type / wake up from potential sleep
+    uint8_t chip_id[3] = {};
+
+    // Temporarily set the flash type to make use of hal_exflash_special_command()
+    flash_type = DEFAULT_SPI_FLASH_TYPE;
+    if (!hal_exflash_special_command(HAL_EXFLASH_SPECIAL_SECTOR_NONE, HAL_EXFLASH_COMMAND_READID, nullptr, chip_id, 0)) {
+        // Try all the supported wake-up command
+        for (auto flash: all_flashes) {
+            flash_type = flash;
+            hal_exflash_special_command(HAL_EXFLASH_SPECIAL_SECTOR_NONE, HAL_EXFLASH_COMMAND_WAKEUP, nullptr, nullptr, 0);
+            nrf_delay_us(20); // Need delay to ensure the flash is woken up.
+        }
+
+        // Read ID must succeed at this point
+        CHECK(hal_exflash_special_command(HAL_EXFLASH_SPECIAL_SECTOR_NONE, HAL_EXFLASH_COMMAND_READID, nullptr, chip_id, 0));
+    }
+
+    flash_type = hal_exflash_get_type(chip_id);
```

Unchecked value, we should probably error out in case of unknown flash type.